cdm/src/main/java/thredds/crawlabledataset/s3/ThreddsS3ClientImpl.java

```diff
@@ -116,4 +123,67 @@ public File saveObjectToFile(S3URI s3uri, File file) throws IOException {
             }
         }
     }
+
+    @Override
+    public ThreddsS3Metadata getMetadata(S3URI s3uri) {
+        ObjectMetadata metadata = getObjectMetadata(s3uri);
+
+        if (metadata != null) {
+            return new ThreddsS3Object(s3uri, metadata.getLastModified(), metadata.getContentLength());
+        } else if (isDirectory(s3uri)) {
+            return new ThreddsS3Directory(s3uri);
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public ThreddsS3Listing listContents(S3URI s3uri) {
+        ThreddsS3Listing listing = new ThreddsS3Listing(s3uri);
+        ObjectListing page = listObjects(s3uri);
+
+        if (page == null) {
+            return null;
+        }
+
+        listing.add(page);
+        int pageNo = 2;
+
+        while (page.isTruncated() && pageNo <= maxListingPages) {
+            page = s3Client.listNextBatchOfObjects(page);
+            logger.info(String.format("Downloaded page %d of S3 listing '%s'", pageNo, s3uri));
+            listing.add(page);
```

Comment:
pageNo is different to the listing size - listing size is not guaranteed per page.